File: precipitation/precipitation_foucaut.py
"""
SDSU Climate Informatics Lab
San Diego State University
by Waverley Moody
Supervised by Distinguished Professor Samuel Shen
Python Code Version 1.0.0

A reproduction of the University of Washington General Circulation
Animations Library by Professor John Michael Wallace.

Script: precipitation_foucaut.py
Description: Generates the precipitation climatology animation from ERA5 reanalysis data (1979-2000), rendered in the Foucaut projection. Weekly totals (cm water equivalent) computed server-side by the derived-era5-single-levels-daily-statistics CDS dataset, averaged into a 22-year climatological mean per 48-frame weekly cycle.
Note: For the Plate Carrée, Robinson, and Nicolosi projections, see the other scripts in the precipitation scripts folder.

--- Implementation notes ---
Uses PROJ's `fouc` operation (plain Foucaut, NOT `fouc_s`/Foucaut
Sinusoidal — a different, parameterized blend family). `fouc` has a
working inverse transform, so unlike Nicolosi this can use a normal
Cartopy Projection subclass and standard GeoAxes rendering rather than a
manual reprojection pipeline.

Foucaut has pointed poles: every longitude at lat=90/-90 collapses to a
single point, so there's no clean row to place longitude labels along
without overlap — longitude labels are omitted, and each pole gets one
label instead of five overlapping ones.
"""

import xarray as xr
import numpy as np
import matplotlib
import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import matplotlib.ticker
import cartopy.crs as ccrs
import cartopy.feature as cfeature
import shapely.geometry as sgeom
import imageio
from pathlib import Path

matplotlib.rcParams['font.family'] = 'Arial'

# ── Paths ────────────────────────────────────────────────────────────────────
DATA_FILE = Path('/Volumes/CLIMATEDATA/precip/climatology/climatology_precip_48frame.nc')
OUTPUT_DIR = Path('/Users/waverleymoody/Downloads/precipitation_animation')
OUTPUT_DIR.mkdir(exist_ok=True, parents=True)


# ── Custom Foucaut projection ──────────────────────────────────────────────────
class Foucaut(ccrs.Projection):
    """PROJ's plain `fouc` operation (NOT `fouc_s`/Foucaut Sinusoidal --
    that's a different, parameterized projection family). Confirmed to
    produce the correct pointed-pole Foucaut shape with no tuning needed.
    """

    def __init__(self):
        super().__init__(proj4_params=[('proj', 'fouc')])

        # Trace the outer boundary: up the antimeridian at lon=-180, then
        # back down at lon=180, forming a closed loop. Vectorized via
        # transform_points rather than a per-point loop.
        lats = np.linspace(-90, 90, 200)
        lons_left = np.full_like(lats, -180)
        lons_right = np.full_like(lats, 180)
        all_lons = np.concatenate([lons_left, lons_right[::-1]])
        all_lats = np.concatenate([lats, lats[::-1]])

        xyz = self.transform_points(ccrs.Geodetic(), all_lons, all_lats)
        x, y = xyz[:, 0], xyz[:, 1]
        valid = np.isfinite(x) & np.isfinite(y)
        x, y = x[valid], y[valid]

        self._boundary = sgeom.LinearRing(zip(x, y))
        # Computed from the actual traced boundary rather than a generic
        # hardcoded guess (the earlier version's hardcoded Robinson-style
        # bounds is what cut off the top/bottom of the map), so this stays
        # correct even if the true extent doesn't match some other
        # projection's typical bounds.
        self._x_limits = (x.min(), x.max())
        self._y_limits = (y.min(), y.max())

    @property
    def boundary(self):
        return self._boundary

    @property
    def x_limits(self):
        return self._x_limits

    @property
    def y_limits(self):
        return self._y_limits

    @property
    def threshold(self):
        return 1e4


# ── Projection to render ──────────────────────────────────────────────────────
PROJ_NAME = 'foucaut'
PROJ_CRS = Foucaut()

# ── Plot settings ─────────────────────────────────────────────────────────────
VMIN, VMAX = 0, 21
TICK_STEP = 3

# Smooth continuous gradient (not discrete flat blocks): white at 0 cm
# gradually blends into blue over just the first 0.3 cm (shrunk from 1 cm
# to cut down on how much of the colorbar reads as blank/white), then
# continues through the standard blue → green → yellow → orange → red →
# purple wet gradient up to 21 cm. Positions are fractions of VMAX.
color_stops = [
    (0 / VMAX, '#ffffff'),    # white           (0 cm)
    (0.3 / VMAX, '#d0e8f0'),  # pale blue       (0.3 cm)
    (2 / VMAX, '#7ab8e0'),    # light blue      (2 cm)
    (3 / VMAX, '#3060c0'),    # blue            (3 cm)
    (6 / VMAX, '#00994d'),    # green           (6 cm)
    (8 / VMAX, '#66cc00'),    # yellow-green    (8 cm)
    (9 / VMAX, '#ffff00'),    # yellow          (9 cm)
    (10 / VMAX, '#ff9900'),   # orange          (10 cm)
    (11 / VMAX, '#ff3300'),   # red-orange      (11 cm)
    (12 / VMAX, '#990000'),   # dark red        (12 cm)
    (21 / VMAX, '#660066'),   # purple          (21 cm)
]
CMAP = mcolors.LinearSegmentedColormap.from_list('custom_precip', color_stops, N=256)
BOUNDS = np.arange(VMIN, VMAX + 1, TICK_STEP)  # 0, 3, 6, ..., up to VMAX
if BOUNDS[-1] != VMAX:
    BOUNDS = np.append(BOUNDS, VMAX)  # always label the true top of the range

# ── Load pre-computed climatology ──────────────────────────────────────────────
ds_clim = xr.open_dataset(DATA_FILE)

# ERA5 stores longitude as 0-360; normalize to -180-180 and sort to prevent
# antimeridian reprojection distortion.
ds_clim = ds_clim.assign_coords(
    longitude=(((ds_clim['longitude'] + 180) % 360) - 180)
).sortby('longitude')

lats = ds_clim['latitude'].values
lons = ds_clim['longitude'].values
labels = ds_clim['frame_label'].values

frames_data = []
for i in range(ds_clim.sizes['frame']):
    field = ds_clim['tp_weekly_total_cm'].isel(frame=i).values
    label = str(labels[i])
    frames_data.append((label, field))
    print(f'Loaded: {label}')

# ── Generate frames + MP4 ───────────────────────────────────────────────────────
print(f'=== Rendering projection: {PROJ_NAME} ===')
frames_dir = OUTPUT_DIR / PROJ_NAME / 'frames'
frames_dir.mkdir(parents=True, exist_ok=True)
frame_paths = []

for i, (label, field) in enumerate(frames_data):
    fig, ax = plt.subplots(
        figsize=(12, 6),
        subplot_kw={'projection': PROJ_CRS}
    )

    im = ax.contourf(
        lons, lats, field,
        levels=np.linspace(VMIN, VMAX, 61),
        cmap=CMAP,
        extend='neither',
        transform=ccrs.PlateCarree()
    )

    ax.add_feature(cfeature.COASTLINE, linewidth=1.0, edgecolor='black')
    ax.add_feature(cfeature.BORDERS, linewidth=0.8, edgecolor='black')
    ax.add_feature(cfeature.STATES, linewidth=0.5, edgecolor='black')
    ax.set_global()

    gl = ax.gridlines(draw_labels=False, linewidth=0.3, color='gray', alpha=0.5)
    gl.xlocator = matplotlib.ticker.FixedLocator([-180, -90, 0, 90, 180])
    gl.ylocator = matplotlib.ticker.FixedLocator([-90, -45, 0, 45, 90])

    # ── Lat/lon labels (projection-aware placement) ──────────────────────────
    # Foucaut has pointed poles: every longitude at lat=90/-90 collapses to
    # the same single point, so there's no clean row to place longitude
    # labels along without them overlapping. Following the confirmed-working
    # reference implementation, longitude labels are omitted; each pole gets
    # a single unambiguous label instead of five overlapping ones.
    x, y = PROJ_CRS.transform_point(0, 90, ccrs.PlateCarree())
    ax.annotate('90°N', xy=(x, y), xycoords='data',
                xytext=(0, 6), textcoords='offset points',
                ha='center', va='bottom', fontsize=9, annotation_clip=False)
    x, y = PROJ_CRS.transform_point(0, -90, ccrs.PlateCarree())
    ax.annotate('90°S', xy=(x, y), xycoords='data',
                xytext=(0, -6), textcoords='offset points',
                ha='center', va='top', fontsize=9, annotation_clip=False)

    # Latitude labels along the left edge — 90°N/90°S are excluded here
    # since they're already covered by the dedicated pole labels above.
    for lat, lat_label in [(-45, '45°S'), (0, '0°'), (45, '45°N')]:
        x, y = PROJ_CRS.transform_point(-180, lat, ccrs.PlateCarree())
        ax.annotate(lat_label, xy=(x, y), xycoords='data',
                    xytext=(-8, 0), textcoords='offset points',
                    ha='right', va='center', fontsize=9, annotation_clip=False)

    # ── Colorbar (rectangular, no pointed extend arrows) ─────────────────────
    cbar = fig.colorbar(im, ax=ax, orientation='horizontal',
                         pad=0.12, fraction=0.05, aspect=20, extend='neither')
    cbar.set_label('Precipitation (cm)', fontsize=11)
    cbar.set_ticks(BOUNDS)
    cbar.set_ticklabels([str(t) for t in BOUNDS])
    cbar.ax.tick_params(labelsize=9)

    # ── Titles ───────────────────────────────────────────────────────────────
    ax.text(0.0, 1.02, 'ERA-5 | Climate Reanalyzer',
             transform=ax.transAxes, fontsize=10,
             fontweight='bold', ha='left', va='bottom')
    ax.text(1.0, 1.02, f'{label}; 1979–2000 Mean Weekly Total',
             transform=ax.transAxes, fontsize=10,
             fontweight='bold', ha='right', va='bottom')

    fpath = frames_dir / f'frame_{i:03d}.png'
    plt.savefig(fpath, dpi=150, bbox_inches='tight')
    plt.close()
    frame_paths.append(str(fpath))
    print(f'Saved frame {i+1}/{len(frames_data)}')

# ── Assemble into MP4 ─────────────────────────────────────────────────────────
with imageio.get_writer(OUTPUT_DIR / 'precipitation_climatology_foucaut.mp4',
                         fps=4, codec='libx264') as writer:
    for fp in frame_paths:
        writer.append_data(imageio.imread(fp))

print('Done! Animation saved to:', OUTPUT_DIR / 'precipitation_climatology_foucaut.mp4')

File: precipitation/precipitation_foucaut.R
# SDSU Climate Informatics Lab
# San Diego State University
# by Waverley Moody
# Supervised by Distinguished Professor Samuel Shen
# R Code Version 1.0.0
#
# A reproduction of the University of Washington General Circulation
# Animations Library by Professor John Michael Wallace.
#
# Script: precipitation_foucaut.R
# Description: Generates the precipitation climatology animation from ERA5 reanalysis data (1979-2000), rendered in the Foucaut projection. Weekly totals (cm water equivalent) computed server-side by the derived-era5-single-levels-daily-statistics CDS dataset, averaged into a 22-year climatological mean per 48-frame weekly cycle.
#
# Note: For the Plate Carrée, Robinson, and Nicolosi projections, see the other scripts in the precipitation scripts folder.

library(terra)
library(ncdf4)
library(rnaturalearth)
library(sf)
library(av)

# ── Paths ────────────────────────────────────────────────────────────────
data_file  <- "/Volumes/CLIMATEDATA/precip/climatology/climatology_precip_48frame.nc"
output_dir <- "/Users/waverleymoody/Downloads/precipitation_animation/r_output"
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

proj_name  <- "foucaut"
target_crs <- "+proj=fouc +datum=WGS84 +units=m +no_defs"

frames_dir <- file.path(output_dir, proj_name, "frames")
dir.create(frames_dir, showWarnings = FALSE, recursive = TRUE)

# ── Plot settings (identical to the PlateCarree/Robinson precipitation
# scripts) ──────────────────────────────────────────────────────────────
vmin <- 0
vmax <- 21
tick_step <- 3

# Same gradient as the Python Plate Carrée/Robinson/Foucaut/Nicolosi
# precipitation scripts: white at 0 cm blends into blue over just the
# first 0.3 cm, then continues through blue -> green -> yellow -> orange
# -> red -> purple up to 21 cm. Positions are fractions of vmax, mirroring
# the Python color_stops list exactly.
precip_colors <- c(
  '#ffffff', # white           (0 cm)
  '#d0e8f0', # pale blue       (0.3 cm)
  '#7ab8e0', # light blue      (2 cm)
  '#3060c0', # blue            (3 cm)
  '#00994d', # green           (6 cm)
  '#66cc00', # yellow-green    (8 cm)
  '#ffff00', # yellow          (9 cm)
  '#ff9900', # orange          (10 cm)
  '#ff3300', # red-orange      (11 cm)
  '#990000', # dark red        (12 cm)
  '#660066'  # purple          (21 cm)
)
precip_positions <- c(0, 0.3, 2, 3, 6, 8, 9, 10, 11, 12, 21) / vmax

# Mirrors LinearSegmentedColormap.from_list with (position, color) tuples,
# since colorRampPalette() only supports evenly-spaced stops -- needed
# here (unlike the temperature Foucaut script) because precipitation's
# color stops are deliberately uneven (white compressed into just the
# first 0.3 cm out of a 21 cm range).
make_custom_palette <- function(colors, positions, n) {
  rgb_mat  <- t(col2rgb(colors)) / 255
  eval_pos <- seq(0, 1, length.out = n)
  r <- approx(positions, rgb_mat[, 1], xout = eval_pos)$y
  g <- approx(positions, rgb_mat[, 2], xout = eval_pos)$y
  b <- approx(positions, rgb_mat[, 3], xout = eval_pos)$y
  rgb(r, g, b)
}

n_colors       <- 60  # matches the Python script's 60 color steps
precip_palette <- make_custom_palette(precip_colors, precip_positions, n_colors)
breaks         <- seq(vmin, vmax, length.out = n_colors + 1)

# ── Natural Earth basemap layers ───────────────────────────────────────────
cat("Downloading/loading basemap layers...\n")
coastline <- ne_coastline(scale = 50, returnclass = "sf")
borders   <- ne_countries(scale = 50, returnclass = "sf")
# US only -- ne_states() with no country filter pulls every country's
# first-level admin boundaries (provinces, states, etc. worldwide), which
# produces dense internal hatching. We only want that level of detail for
# the US.
states    <- ne_states(country = "United States of America", returnclass = "sf")

# Reproject once, not per frame
coastline_p <- st_transform(coastline, target_crs)
borders_p   <- st_transform(borders, target_crs)
states_p    <- st_transform(states, target_crs)

# ── Load pre-computed climatology ──────────────────────────────────────────
r_stack <- rast(data_file, subds = "tp_weekly_total_cm")  # already in cm, no conversion needed
r_stack <- rotate(r_stack)              # ERA5 is 0-360 -- fix before reprojecting
n_frames <- nlyr(r_stack)

nc <- nc_open(data_file)
frame_labels <- ncvar_get(nc, "frame_label")
nc_close(nc)

cat(sprintf("Loaded %d frames\n", n_frames))

# ── Helper: project a single lon/lat point into a target CRS ───────────────
proj_point <- function(lon, lat, target_crs) {
  pts <- vect(cbind(lon, lat), crs = "EPSG:4326")
  pts_proj <- project(pts, target_crs)
  crds(pts_proj)
}

# ── Helper: build a graticule line (meridian or parallel) in a target CRS ──
make_meridian <- function(lon, target_crs) {
  lats <- seq(-89.9, 89.9, length.out = 100)
  ln <- st_linestring(cbind(lon, lats))
  st_transform(st_sfc(ln, crs = "EPSG:4326"), target_crs)
}
make_parallel <- function(lat, target_crs) {
  lons <- seq(-180, 180, length.out = 200)
  ln <- st_linestring(cbind(lons, lat))
  st_transform(st_sfc(ln, crs = "EPSG:4326"), target_crs)
}

meridians <- lapply(c(-180, -90, 0, 90, 180), make_meridian, target_crs = target_crs)
parallels <- lapply(c(-90, -45, 0, 45, 90), make_parallel, target_crs = target_crs)

# ── Build clipping boundary (the true Foucaut diamond outline) ─────────────
# terra::project() on a raster fills the full rectangular bounding box and
# inverse-projects each cell back to a source lon/lat; outside the true
# projection domain that comes back wrapped/extrapolated rather than NA,
# producing a striped background outside the continents. This traces the
# up-the-antimeridian-and-back outline and masks the raster to it after
# projecting, matching the Python version's boundary handling.
lats_b    <- seq(-90, 90, length.out = 200)
left_pts  <- vect(cbind(-180, lats_b), crs = "EPSG:4326")
right_pts <- vect(cbind(180, rev(lats_b)), crs = "EPSG:4326")
boundary_pts    <- rbind(project(left_pts, target_crs), project(right_pts, target_crs))
boundary_coords <- crds(boundary_pts)
boundary_coords <- rbind(boundary_coords, boundary_coords[1, ])  # close the ring
boundary_poly   <- vect(boundary_coords, type = "polygons", crs = target_crs)

# ── Generate frames + MP4 ───────────────────────────────────────────────────
frame_paths <- character(n_frames)

for (i in seq_len(n_frames)) {
  
  field   <- r_stack[[i]]
  field_p <- mask(project(field, target_crs), boundary_poly)
  
  label <- frame_labels[i]
  fpath <- file.path(frames_dir, sprintf("frame_%03d.png", i - 1))
  
  png(fpath, width = 1800, height = 980, res = 150)
  layout(matrix(1:2, nrow = 2), heights = c(5, 1.3))
  
  # ── Map panel ──
  # bty = "n" reliably suppresses the plot-region box; box=FALSE isn't a
  # real terra::plot() parameter and gets silently dropped.
  par(mar = c(2, 3, 7.5, 2), bty = "n")
  plot(field_p, col = precip_palette, breaks = breaks, range = c(vmin, vmax),
       axes = FALSE, legend = FALSE, main = "")
  
  plot(st_geometry(coastline_p), add = TRUE, lwd = 1.0, col = "black")
  plot(st_geometry(borders_p), add = TRUE, lwd = 0.8, border = "black")
  plot(st_geometry(states_p), add = TRUE, lwd = 0.5, border = "black")
  
  for (m in meridians) plot(m, add = TRUE, col = "gray", lwd = 0.3)
  for (p in parallels) plot(p, add = TRUE, col = "gray", lwd = 0.3)
  
  # Pole labels -- Foucaut's poles are single points, not an edge to walk
  # longitudes along like Robinson/PlateCarree, so these are two fixed
  # annotations rather than a loop.
  xy_n <- proj_point(0, 90, target_crs)
  text(xy_n[1], xy_n[2], labels = "90\u00b0N", pos = 3, cex = 0.95,
       xpd = NA, offset = 0.3)
  xy_s <- proj_point(0, -90, target_crs)
  text(xy_s[1], xy_s[2], labels = "90\u00b0S", pos = 1, cex = 0.95,
       xpd = NA, offset = 0.3)
  
  # Latitude labels along the left edge only. No longitude labels --
  # Foucaut's pointed poles make bottom-edge longitude labels overlap into
  # unreadable text, same reasoning as the Python version.
  lat_labels <- list(c(-45, "45\u00b0S"), c(0, "0\u00b0"), c(45, "45\u00b0N"))
  for (ll in lat_labels) {
    lat_val <- as.numeric(ll[1]); lab <- ll[2]
    xy <- proj_point(-180, lat_val, target_crs)
    text(xy[1], xy[2], labels = lab, pos = 2, cex = 0.95, xpd = NA, offset = 0.4)
  }
  
  mtext("ERA-5 | Climate Reanalyzer", side = 3, line = 6, adj = 0,
        cex = 1.0, font = 2)
  mtext(sprintf("%s; 1979\u20132000 Weekly Total Mean", label), side = 3, line = 6,
        adj = 1, cex = 1.0, font = 2)
  
  # ── Colorbar panel ──
  # Custom-drawn via image() rather than terra's plot(..., legend.only=TRUE):
  # that legend mechanism positions itself in whole-device NDC coordinates,
  # not panel-local ones, so it doesn't respect this layout panel. Drawing
  # it directly with image() gives full control and keeps it inside this
  # panel.
  par(mar = c(4.5, 6, 1.2, 6), bty = "n")
  cbar_mat <- matrix(seq(vmin, vmax, length.out = n_colors), ncol = 1)
  # xlim wider than the actual vmin/vmax data range lSeaves blank space on
  # both sides within the panel, so the bar reads as shorter/narrower
  # instead of stretching edge-to-edge across the full frame width.
  #
  # x = breaks (not seq(vmin, vmax, length.out=n_colors)) is important:
  # passing n_colors points makes image() treat them as bin CENTERS, which
  # auto-extends the painted area half a bin-width past both vmin and
  # vmax -- that's what let the purple bleed past the vmax=21 border in
  # the first render. Passing the n_colors+1 breaks instead makes x the
  # true cell edges, so the painted region stops exactly at vmax.
  pad <- (vmax - vmin) * 0.25
  image(x = breaks, y = 1, z = cbar_mat,
        col = precip_palette, breaks = breaks,
        axes = FALSE, xlab = "", ylab = "",
        xlim = c(vmin - pad, vmax + pad))
  usr <- par("usr")
  rect(vmin, usr[3], vmax, usr[4], border = "black", lwd = 1.2)
  axis(1, at = seq(vmin, vmax, tick_step), labels = seq(vmin, vmax, tick_step),
       cex.axis = 1.0)
  mtext("Precipitation (cm)", side = 1, line = 3.0, cex = 1.0)
  
  dev.off()
  
  frame_paths[i] <- fpath
  cat(sprintf("Saved frame %d/%d\n", i, n_frames))
}

video_path <- file.path(output_dir, "precipitation_climatology_foucaut.mp4")
av_encode_video(frame_paths, output = video_path, framerate = 4, codec = "libx264")

cat(sprintf("Done! Animation saved to: %s\n", video_path))